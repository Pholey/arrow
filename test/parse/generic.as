
// Declare: Generic Record
struct Box<T> { value: T; }
struct Point<T> { x: T; y: T; }
struct Pair<T, U> { first: T; second: U; }

// Declare: Generic Function
def method<T>(param: T): T {
  return param;
}

// Use: Generic Record
let a: Box<int32>;
let b: Pair<Box<int32>, Box<bool>>;

// Use: Generic Method
method<int64>(302);
method<Box<int32>>(a);
method<Pair<Box<int32>, Box<bool>>>(b);
