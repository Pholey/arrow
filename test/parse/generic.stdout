{
  "tag": "Module",
  "span": "test/parse/generic.as:1-19",
  "statements": [
    {
      "tag": "TypeRecord",
      "span": "test/parse/generic.as:3:1-28",
      "name": "Box",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:3:12-13",
          "name": "T"
        }
      ],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/generic.as:3:17-25",
          "name": "value",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:3:24-25",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:3:24-25",
              "text": "T"
            }
          }
        }
      ]
    },
    {
      "tag": "TypeRecord",
      "span": "test/parse/generic.as:4:1-32",
      "name": "Point",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:4:14-15",
          "name": "T"
        }
      ],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/generic.as:4:19-23",
          "name": "x",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:4:22-23",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:4:22-23",
              "text": "T"
            }
          }
        },
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/generic.as:4:25-29",
          "name": "y",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:4:28-29",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:4:28-29",
              "text": "T"
            }
          }
        }
      ]
    },
    {
      "tag": "TypeRecord",
      "span": "test/parse/generic.as:5:1-43",
      "name": "Pair",
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:5:13-14",
          "name": "T"
        },
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:5:16-17",
          "name": "U"
        }
      ],
      "members": [
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/generic.as:5:21-29",
          "name": "first",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:5:28-29",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:5:28-29",
              "text": "T"
            }
          }
        },
        {
          "tag": "TypeRecordMember",
          "span": "test/parse/generic.as:5:31-40",
          "name": "second",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:5:39-40",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:5:39-40",
              "text": "U"
            }
          }
        }
      ]
    },
    {
      "tag": "Function",
      "span": "test/parse/generic.as:8-10",
      "name": "method",
      "result_type": {
        "tag": "TypeName",
        "span": "test/parse/generic.as:8:26-27",
        "name": {
          "tag": "Identifier",
          "span": "test/parse/generic.as:8:26-27",
          "text": "T"
        }
      },
      "parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:8:15-23",
          "name": "param",
          "type": {
            "tag": "TypeName",
            "span": "test/parse/generic.as:8:22-23",
            "name": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:8:22-23",
              "text": "T"
            }
          }
        }
      ],
      "type_parameters": [
        {
          "tag": "Parameter",
          "span": "test/parse/generic.as:8:12-13",
          "name": "T"
        }
      ],
      "block": {
        "tag": "Block",
        "span": "test/parse/generic.as:8-10",
        "statements": [
          {
            "tag": "Return",
            "span": "test/parse/generic.as:9:3-15",
            "operand": {
              "tag": "Identifier",
              "span": "test/parse/generic.as:9:10-15",
              "text": "param"
            }
          }
        ]
      }
    },
    {
      "tag": "Variable",
      "span": "test/parse/generic.as:13:1-19",
      "name": "a",
      "mutable": false,
      "type": {
        "tag": "TypeName",
        "span": "test/parse/generic.as:13:8-18",
        "name": {
          "tag": "Identifier",
          "span": "test/parse/generic.as:13:8-18",
          "text": "Box"
        }
      },
      "initializer": null
    },
    {
      "tag": "Variable",
      "span": "test/parse/generic.as:14:1-36",
      "name": "b",
      "mutable": false,
      "type": {
        "tag": "TypeName",
        "span": "test/parse/generic.as:14:8-35",
        "name": {
          "tag": "Identifier",
          "span": "test/parse/generic.as:14:8-35",
          "text": "Pair"
        }
      },
      "initializer": null
    },
    {
      "tag": "Call",
      "span": "test/parse/generic.as:17:14-19",
      "operand": {
        "tag": "Identifier",
        "span": "test/parse/generic.as:17:1-14",
        "text": "method"
      },
      "arguments": [
        {
          "tag": "Argument",
          "span": "test/parse/generic.as:17:15-18",
          "value": {
            "tag": "Integer",
            "span": "test/parse/generic.as:17:15-18",
            "value": "302"
          }
        }
      ]
    },
    {
      "tag": "Call",
      "span": "test/parse/generic.as:18:19-22",
      "operand": {
        "tag": "Identifier",
        "span": "test/parse/generic.as:18:1-19",
        "text": "method"
      },
      "arguments": [
        {
          "tag": "Argument",
          "span": "test/parse/generic.as:18:20-21",
          "value": {
            "tag": "Identifier",
            "span": "test/parse/generic.as:18:20-21",
            "text": "a"
          }
        }
      ]
    },
    {
      "tag": "Call",
      "span": "test/parse/generic.as:19:36-39",
      "operand": {
        "tag": "Identifier",
        "span": "test/parse/generic.as:19:1-36",
        "text": "method"
      },
      "arguments": [
        {
          "tag": "Argument",
          "span": "test/parse/generic.as:19:37-38",
          "value": {
            "tag": "Identifier",
            "span": "test/parse/generic.as:19:37-38",
            "text": "b"
          }
        }
      ]
    }
  ]
}
